exo-go/src/app_runner/index.go

```diff
@@ -0,0 +1,145 @@
+			appRunner.Env = append(appRunner.Env, fmt.Sprintf("%s=%s", variable, value))
+		}
+	}
+	return appRunner
+}
+
+// Start runs the application
+func (appRunner *AppRunner) Start() {
+	_, stdoutBuffer, err := dockerCompose.RunAllImages(appRunner.Env, appRunner.DockerConfigLocation, appRunner.Write)
+	if err != nil {
+		appRunner.Shutdown("", "Failed to run images")
+	} else {
+		appRunner.compileOnlineText(func(err error) {
+			if err != nil {
+				log.Fatal(err)
+			} else {
+				wg := new(sync.WaitGroup)
+				for role, onlineText := range appRunner.OnlineTexts {
+					wg.Add(1)
+					go func(role string, stdoutBuffer fmt.Stringer, onlineText string) {
+						defer wg.Done()
+						processHelpers.Wait(stdoutBuffer, onlineText, func() {
+							appRunner.Logger.Log(role, fmt.Sprintf("'%s' is running", role), true)
+						})
+					}(role, stdoutBuffer, onlineText)
+				}
+				wg.Wait()
+				appRunner.Write("all services online")
+			}
+		})
+	}
+}
+
+// Shutdown shuts down the application
+func (appRunner *AppRunner) Shutdown(closeMessage, errorMessage string) {
+	var exitCode int
+	if len(errorMessage) > 0 {
+		color.Red(errorMessage)
+		exitCode = 1
+	} else {
+		fmt.Printf("\n\n%s", closeMessage)
+		exitCode = 0
+	}
+	cmd, stdoutBuffer, err := dockerCompose.KillAllContainers([]string{}, appRunner.DockerConfigLocation, appRunner.Write)
+	if err != nil {
+		log.Fatalf("Failed to start killing all containers.\nOutput:%s\nError:%s", stdoutBuffer.String(), err)
+	}
+	if err := cmd.Wait(); err != nil {
+		log.Fatalf("Failed to kill all containers.\nOutput:%s\nError:%s", stdoutBuffer.String(), err)
+	} else {
+		os.Exit(exitCode)
+	}
+}
+
+// Write logs exo-run output
+func (appRunner *AppRunner) Write(text string) {
+	appRunner.Logger.Log("exo-run", text, true)
+}
+
+func (appRunner *AppRunner) compileOnlineText(done func(error)) {
+	appRunner.OnlineTexts = make(map[string]string)
+	for _, dependency := range appRunner.AppConfig.Dependencies {
+		appRunner.OnlineTexts[dependency.Name] = dependency.GetOnlineText()
+	}
+	wg := new(sync.WaitGroup)
+	for service, serviceData := range serviceHelpers.GetServiceConfigs(appRunner.AppConfig.Services) {
+		wg.Add(1)
+		go func(service string, serviceData types.ServiceConfig) {
+			defer wg.Done()
+			appRunner.getOnlineText(service, serviceData, func(err error) {
+				if err != nil {
+					done(err)
+				}
+			})
+		}(service, serviceData)
+	}
+	wg.Wait()
+	done(nil)
+}
+
+func (appRunner *AppRunner) getOnlineText(role string, serviceData types.ServiceConfig, done func(error)) {
+	if len(serviceData.Location) > 0 {
+		serviceConfig := serviceConfigHelpers.GetServiceConfig(serviceData.Location)
+		appRunner.OnlineTexts[role] = serviceConfig.Startup["online-text"]
+		done(nil)
+	} else if len(serviceData.DockerImage) > 0 {
+		c, err := client.NewEnvClient()
+		if err != nil {
+			panic(err)
+		}
+		dockerHelpers.CatFile(c, serviceData.DockerImage, "service.yml", func(err error, yamlFile []byte) {
```

Comment:
In Go there are no async functions. Functions are synchronous and block until they get their results. If you want to wait for a result that happens later, use channels.